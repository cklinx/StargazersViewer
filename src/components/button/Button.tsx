import React, {useState} from 'react';
import {Pressable as RNPressable, PressableProps, ViewStyle, Animated} from 'react-native';

const AnimatedPressable = Animated.createAnimatedComponent(RNPressable);

interface Props extends PressableProps {
  children?: React.ReactNode;
  scaleValue?: number;
  friction?: number;
  tension?: number;
  style?: ViewStyle | ViewStyle[];
}

export const Button = ({children, scaleValue, friction, tension, style, ...restProps}: Props) => {
  const [scale] = useState(new Animated.Value(1));

  const animatedStyles: ViewStyle = {
    transform: [
      {
        //@ts-ignore
        scale,
      },
    ],
  };

  const shrinkView = () => {
    Animated.spring(scale, {
      toValue: scaleValue ? scaleValue : 0.9,
      useNativeDriver: true,
      friction: 7,
      tension: 40,
    }).start();
  };

  const scaleView = () => {
    Animated.spring(scale, {
      toValue: 1,
      useNativeDriver: true,
      friction: friction ? friction : 7,
      tension: tension ? tension : 40,
    }).start();
  };

  return (
    <AnimatedPressable
      {...restProps}
      style={[animatedStyles, style]}
      onPressIn={() => shrinkView()}
      onPressOut={() => scaleView()}>
      {children}
    </AnimatedPressable>
  );
};
