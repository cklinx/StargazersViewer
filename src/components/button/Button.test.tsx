import React from 'react';
import {Text} from 'react-native';
import {fireEvent, render} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {Button} from './Button';
const mockLabel = 'Test label';
const mockOnPress = jest.fn();
const setup = (overrideOrAddProps?: Partial<React.ComponentProps<typeof Button>>) =>
  render(<Button accessibilityLabel={mockLabel} onPress={mockOnPress} {...overrideOrAddProps} />);

describe('Button', () => {
  it('should render correctly', () => {
    const {toJSON} = renderer.create(
      <Button>
        <Text>text</Text>
      </Button>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should handle being pressed', () => {
    const {getByA11yLabel} = setup();

    fireEvent.press(getByA11yLabel(mockLabel));

    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });
});
