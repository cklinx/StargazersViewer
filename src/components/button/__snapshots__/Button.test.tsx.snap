// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text>
    text
  </Text>
</View>
`;
